server/entity/splash_potion.go

```diff
@@ -0,0 +1,248 @@
+	AddEffect(e effect.Effect)
+}
+
+// Tick ...
+func (s *SplashPotion) Tick(current int64) {
+	if s.closeNextTick {
+		_ = s.Close()
+		return
+	}
+	s.mu.Lock()
+	m, result := s.c.TickMovement(s, s.pos, s.vel, s.yaw, s.pitch, s.ignores)
+	s.pos, s.vel, s.yaw, s.pitch = m.pos, m.vel, m.yaw, m.pitch
+	s.mu.Unlock()
+
+	s.ticksLived++
+	m.Send()
+
+	if m.pos[1] < cube.MinY && current%10 == 0 {
+		s.closeNextTick = true
+		return
+	}
+
+	if result != nil {
+		w := s.World()
+		aabb := s.AABB().Translate(m.pos)
+
+		effects := s.t.Effects()
+		hasEffects := len(effects) > 0
+
+		colour := color.RGBA{R: 0x38, G: 0x5d, B: 0xc6, A: 0xff}
+
+		if hasEffects {
+			colour, _ = effect.ResultingColour(effects)
+
+			ignore := func(entity world.Entity) bool {
+				_, canSplash := entity.(splashable)
+				return !canSplash || entity == s
+			}
+
+			for _, otherEntity := range w.EntitiesWithin(aabb.GrowVec3(mgl64.Vec3{8.25, 4.25, 8.25}), ignore) {
+				splashEntity := otherEntity.(splashable)
+				if !splashEntity.AABB().Translate(splashEntity.Position()).IntersectsWith(aabb.GrowVec3(mgl64.Vec3{4.125, 2.125, 4.125})) {
+					continue
+				}
+
+				distance := world.Distance(EyePosition(splashEntity), m.pos)
+				if distance > 4 {
+					continue
+				}
+
+				distanceMultiplier := 1 - distance/4
+				if entityResult, ok := result.(trace.EntityResult); ok && entityResult.Entity() == otherEntity {
+					distanceMultiplier = 1
+				}
+
+				for _, eff := range effects {
+					if _, ok := eff.Type().(effect.LastingType); !ok {
+						splashEntity.AddEffect(eff.WithPotency(distanceMultiplier))
+						continue
+					}
+
+					distanceAccountedDuration := time.Duration(float64(eff.Duration().Milliseconds())*0.75*distanceMultiplier) * time.Millisecond
+					if distanceAccountedDuration < time.Second {
+						continue
+					}
+					splashEntity.AddEffect(eff.WithDuration(distanceAccountedDuration))
+				}
+			}
+		} else if blockResult, ok := result.(trace.BlockResult); ok && s.t.Equals(potion.Water()) {
+			blockPos := blockResult.BlockPosition().Side(blockResult.Face())
+			if w.Block(blockPos) == fire() {
+				w.SetBlock(blockPos, air())
+			}
+
+			for _, f := range cube.HorizontalFaces() {
+				horizontalPos := blockPos.Side(f)
+				if w.Block(horizontalPos) == fire() {
+					w.SetBlock(horizontalPos, air())
+				}
+			}
+		}
+
+		w.AddParticle(m.pos, particle.Splash{Colour: colour})
+		w.PlaySound(m.pos, sound.GlassBreak{})
+
+		s.closeNextTick = true
+	}
+}
+
+// ignores returns whether the SplashPotion should ignore collision with the entity passed.
+func (s *SplashPotion) ignores(entity world.Entity) bool {
+	_, ok := entity.(Living)
+	return !ok || entity == s || (s.ticksLived < 5 && entity == s.owner)
+}
+
+// New creates a SplashPotion with the position, velocity, yaw, and pitch provided. It doesn't spawn the SplashPotion,
+// only returns it.
+func (s *SplashPotion) New(pos, vel mgl64.Vec3, yaw, pitch float64) world.Entity {
```

Comment:
Would accept the potion type here too. No need to implement the Projectile interface, I removed that in the latest commit. Make sure to pull that commit though.